### ISSUE TITLE
Failure - repeatedly sample same index

### ISSUE THREAD

**Issue by JohanYe** [leonardodalinky/fpsample#2]
First of all, thanks for your great work! I use your method inside a pytorch dataloader as subsampling for my point clouds. This frequently leads to errors (for some datasets it's around 60% i'd say). 

The issue is that the indices sampled using fpsampling, ends up repeating the same point many times, despite there being enough indices to choose from. In the attached picture, you see that the index 0 has been sampled repeatedly. This usually happens by the method sampling index 0 repeatedly, but it can also happen to other indices.
![image](https://github.com/user-attachments/assets/074f2118-62b6-4374-b582-69e44cc70db0)

I thought it was a bad seed, so i selected starting_idx to be able to reproduce this error, but it sometimes disappears, despite using the exact same starting_idx in the picture below,and running the same code from the previous image (so same point cloud). It suddenly samples the point cloud correctly
![image](https://github.com/user-attachments/assets/a2477b38-5324-4016-9f1e-3b8c158a1ff6)

**Comment by leonardodalinky** [comment]
Thank you for your report. I think there should be more information to solve this problem. Could you somehow show examples of the raw data when this issue occurs? It is hard to debug this without reproduction.

Two more questions are also helpful:
* Does this issue still exists in other backend implementation, such as `bucket_fps_kdtree_sampling` and `fps_npdu_kdtree_sampling`?
* Does your data have repeated points?

**Comment by JohanYe** [comment]
Thanks for the quick reply!

A picture of the input point cloud
![image](https://github.com/user-attachments/assets/181e35b7-c2fb-4927-be5b-e57ab00081c9)

I don't believe there are any repeated points:
![image](https://github.com/user-attachments/assets/d726b457-916a-4718-9c37-6ac65a9c49d6)

It occurs in `bucket_fps_kdtree_sampling`, but not `fps_npdu_kdtree_sampling`'.

I have uploaded the mesh i am testing on here: https://www.dropbox.com/scl/fi/7gsbyad87xzge57sn1b4j/02433_upper_unn3_mesh.ply?rlkey=j9l1fqkhkushoxoe3au0063qv&dl=0.

**Comment by leonardodalinky** [comment]
I would take some time to find out the source of this issue.

But the most weird thing is, are you sure that the bucket-based sampling works occasionally even for **the same point cloud with the same parameters**?

Also, I notice that when the sampling functions work in some cases, the first index of the sampled point is incorrect.

> I thought it was a bad seed, so i selected starting_idx to be able to reproduce this error, but it sometimes disappears, despite using the exact same starting_idx in the picture below,and running the same code from the previous image (so same point cloud). It suddenly samples the point cloud correctly ![image](https://private-user-images.githubusercontent.com/57265800/363280117-a2477b38-5324-4016-9f1e-3b8c158a1ff6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUxMDk1NTYsIm5iZiI6MTcyNTEwOTI1NiwicGF0aCI6Ii81NzI2NTgwMC8zNjMyODAxMTctYTI0NzdiMzgtNTMyNC00MDE2LTlmMWUtM2I4YzE1OGExZmY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODMxVDEzMDA1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkYmRlZjRjMjhlNzk1ODcyNDZjYWIyZmE0Y2U2N2Q0NTMyYWJjNDNlY2MxZTNjZmE2NzcxNTA2N2MyMGY3ZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.URNUQXc1Sr_h3ou20TSzeBNAOEA4uOpbu3QUthvaWA8)

I am pretty sure that the index of first sampled should be the `start_idx`. Is this always the case?

**Comment by leonardodalinky** [comment]
After I try to reproduce this issue, all the sampling functions work well in my Jupyter Notebook

```python
import numpy as np
import fpsample
from plyfile import PlyData

> plydata = PlyData.read("02433_upper_unn3_mesh.ply")

> datum = np.stack([plydata["vertex"]["x"], plydata["vertex"]["y"], plydata["vertex"]["z"]], axis=1)
> datum.shape
(3001, 3)

> fpsample.bucket_fps_kdline_sampling(datum, 2048, h=3, start_idx=2)
array([   2, 1312, 2657, ..., 2099,  578,  274], dtype=uint64)
```

You should give a full description of the runtime environment. Parallelism and unexpected memory corruption may result in this issue.

**Comment by JohanYe** [comment]
Since you asked, i tried to call it in the main script outside the dataloader. i managed to produce the error. 
![image](https://github.com/user-attachments/assets/782efecd-bcea-424a-b1af-1a10fb2a64dd)

And it seems like you are right. When the algorithm "works", it still doesn't sample the fulll amount
![image](https://github.com/user-attachments/assets/4aee8417-a552-4fe4-b914-fac8e8bb3231)
![image](https://github.com/user-attachments/assets/0686143f-378b-4d26-91d4-b2405d04fda1)


I have seen the error across different environments, but i have attached an environment where i observe the error. I am running on ubuntu 22.04. (i had to convert my .yml file to txt to upload it to github)
[environment.txt](https://github.com/user-attachments/files/16825479/environment.txt)

**Comment by JohanYe** [comment]
Inspired by your message, i created a jupyter notebook to reproduce the error. 
![image](https://github.com/user-attachments/assets/6a3ee410-49e5-4ef7-92fb-e6562150e67d)

but i also don't see a clear pattern of how many samples are produced, when the method doesn't start from the correct index. Although, it for reason almost always chooses to start from indices 2974 or 2174 or 3000. I have not seen other starting indices

![image](https://github.com/user-attachments/assets/eb7f34c7-13b3-406c-b4a3-3d6810d80895)
![image](https://github.com/user-attachments/assets/be2d27d2-8116-41a5-b192-6f26a7ea873a)
![image](https://github.com/user-attachments/assets/610be288-32c6-4153-96b2-0311eb85dcaa)

**Comment by leonardodalinky** [comment]
That's really weird and unexplainable. Would you mind testing this using [torch version of fpsample](https://github.com/leonardodalinky/pytorch_fpsample)? Though this library is half-done, the CPU mode should work now.

**Comment by leonardodalinky** [comment]
OK finally I reproduce this error, probably caused by the memory layout of the tensors.

For example:
```python
# This works
datum = np.stack([plydata["vertex"]["x"], plydata["vertex"]["y"], plydata["vertex"]["z"]], axis=1)

# This fails!!!
datum = np.stack([plydata.elements[0]["x"], plydata.elements[0]["y"], plydata.elements[0]["z"]])
datum = datum.T
```

I would try to fix this ASAP. Thank you for your response! 👍

**Comment by JohanYe** [comment]
Thanks for the quick support and help! I'm glad you found it relatively quick!

**Comment by leonardodalinky** [comment]
Try v0.3.3 later. Thanks!
